### ISSUE TITLE
[Feature] Implement crypto.createCipheriv

### ISSUE THREAD

**Issue by vicb** [cloudflare/workerd#3277]
`crypto.createCipheriv` is required by Next Auth.

See https://github.com/opennextjs/opennextjs-cloudflare/issues/206

Can we prioritize the implementation of the API?

@jasnell 

The crypto APIs are used by Next from [`crypto-utils.ts`](https://github.com/vercel/next.js/blob/canary/packages/next/src/server/crypto-utils.ts)

What is used by Next is:

// * Ciphers
//   * [x] crypto.createCipheriv(algorithm, key, iv[, options])
//   * [x] crypto.createDecipheriv(algorithm, key, iv[, options])

with the `aes-256-gcm` algo

**Comment by jasnell** [comment]
Was can likely prioritize more of the `node:crypto` APIs in 1Q

**Comment by albertocavalcante** [comment]
Hi there! Do you know if there's any workaround in the meantime? Thanks!

**Comment by jasnell** [comment]
The only workaround is to use web crypto.

**Comment by vicb** [comment]
> The only workaround is to use web crypto.

The code is the Nextjs lib, not in user code so changing to use webcrypto is not trivial.
I would way your only chance is to check if you can setup Next to use a different algo.

**Comment by Macil** [comment]
Small correction to the original post: in the issue https://github.com/opennextjs/opennextjs-cloudflare/issues/206 using the package next-auth, the relevant call to `crypto.createCipheriv` actually happens with the `"aes-256-cbc"` algorithm (through [next-auth/packages/core/src/jwt.ts](https://github.com/nextauthjs/next-auth/blob/6a72f3d041a0e4ac0d311b59f6c5f1e969f2cd5a/packages/core/src/jwt.ts#L56) which eventually calls into [jose/src/runtime/node/encrypt.ts](https://github.com/panva/jose/blob/08be0340ea04eef260bfc206a7962011d1f9ba14/src/runtime/node/encrypt.ts#L39)).

**Comment by vicb** [comment]
> Small correction to the original post: in the issue [opennextjs/opennextjs-cloudflare#206](https://github.com/opennextjs/opennextjs-cloudflare/issues/206) using the package next-auth, the relevant call to `crypto.createCipheriv` actually happens with the `"aes-256-cbc"` algorithm (through [next-auth/packages/core/src/jwt.ts](https://github.com/nextauthjs/next-auth/blob/6a72f3d041a0e4ac0d311b59f6c5f1e969f2cd5a/packages/core/src/jwt.ts#L56) which eventually calls into [jose/src/runtime/node/encrypt.ts](https://github.com/panva/jose/blob/08be0340ea04eef260bfc206a7962011d1f9ba14/src/runtime/node/encrypt.ts#L39)).

I'm pretty sure I tested it with Next 15, any chance you are using a different version than the one of the original bug report?

Anyway we should still implement the algo.

Thanks for your message.

**Comment by Macil** [comment]
Oh sorry I was very zoned into thinking about my specific case while writing that out the first time, and I'm definitely not sure now that spot of code is the only relevant one, especially considering different versions and configurations. I've added a reproduction case on that original issue thread now.

**Comment by arnavgupta00** [comment]
> Hi there! Do you know if there's any workaround in the meantime? Thanks!

I created this [Repo] (https://github.com/arnavgupta00/deployment-cf-workers-prisma-nextauth), This approach addresses the crypto.createCipheriv limitation by substituting Node.js's crypto module with the Web Crypto API

Check it out, might be of help

**Comment by vicb** [comment]
> > Hi there! Do you know if there's any workaround in the meantime? Thanks!
> 
> I created this [Repo](https://github.com/arnavgupta00/deployment-cf-workers-prisma-nextauth), This approach addresses the crypto.createCipheriv limitation by substituting Node.js's crypto module with the Web Crypto API
> 
> Check it out, might be of help

Thanks @arnavgupta00!

If I understand correctly, the main point is to override the encode/decode methods here:

https://github.com/arnavgupta00/deployment-cf-workers-prisma-nextauth/blob/32aa1768edfef8ea112cd5065e2e9c07d5506866/src/lib/authConfig.ts#L43-L46

**Comment by arnavgupta00** [comment]
> > > Hi there! Do you know if there's any workaround in the meantime? Thanks!
> > 
> > 
> > I created this [Repo](https://github.com/arnavgupta00/deployment-cf-workers-prisma-nextauth), This approach addresses the crypto.createCipheriv limitation by substituting Node.js's crypto module with the Web Crypto API
> > Check it out, might be of help
> 
> Thanks [@arnavgupta00](https://github.com/arnavgupta00)!
> 
> If I understand correctly, the main point is to override the encode/decode methods here:
> 
> https://github.com/arnavgupta00/deployment-cf-workers-prisma-nextauth/blob/32aa1768edfef8ea112cd5065e2e9c07d5506866/src/lib/authConfig.ts#L43-L46



Yes, that's correct, I've overridden the default `encode` and `decode`  of next-auth . By implementing custom `encode` and `decode` functions, we can handle JWT operations securely within the Cloudflare environment.

**Comment by vicb** [comment]
Thanks @arnavgupta00, do you know if this would work the same with Next15? (I'm planning to add some docs until the workerd implementation is ready)

**Comment by vicb** [comment]
> Yes, based on the Next.js 15

Thanks!

[This line](https://github.com/arnavgupta00/deployment-cf-workers-prisma-nextauth/blob/32aa1768edfef8ea112cd5065e2e9c07d5506866/src/lib/webcrypto.ts#L4C1-L4C44) looks fishy as `process.env` is not always populated at top level - only guaranteed in the request context. Ping me on Discord to discuss more so that we don't add more OpenNext related messages to this thread, thanks (or on the OpenNext issue)

**Comment by nathan-scheele** [comment]
> > Hi there! Do you know if there's any workaround in the meantime? Thanks!
> 
> I created this [Repo](https://github.com/arnavgupta00/deployment-cf-workers-prisma-nextauth), This approach addresses the crypto.createCipheriv limitation by substituting Node.js's crypto module with the Web Crypto API
> 
> Check it out, might be of help

@arnavgupta00 Does your implementation works with any of [Next-Auth's built-in OAuth providers](https://next-auth.js.org/providers/)?

I have run into the same `crypto.createCipheriv is not implemented yet!` issue, but using [next-authv5](https://authjs.dev/getting-started/migrating-to-v5) and trying to authenticate with Google, as opposed to v4 from your example. It seems to me, in v5 at least, that  `crypto.createCipheriv` is being called by the OAuth providers themselves. 

Curious to know if there's a workaround here.

**Comment by sassanh** [comment]
I'm also having this issue with next-auth v5. It is the only blocking issue for me to migrate from next-on-pages to opennextjs.
It would be really nice if we had an ETA for the implementation of createCipheriv so that we can plan whether we should get back to next-on-pages or we can rely on opennextjs to work for us in the near future.

**Comment by jasnell** [comment]
Over the past few days I've been working on the implementation of `createPrivateKey(...)`, `createPublicKey(...)`, `generateKey(...)`, etc. Early this next week I'll work on the implementation of `generateKeyPair(...)`. Once the key creation/generation functions are complete, I'll be immediately turning attention to the cipher/decipher and sign/verify APIs. Making progress even if slowly.

**Comment by arnavgupta00** [comment]
> > > Hi there! Do you know if there's any workaround in the meantime? Thanks!
> > 
> > 
> > I created this [Repo](https://github.com/arnavgupta00/deployment-cf-workers-prisma-nextauth), This approach addresses the crypto.createCipheriv limitation by substituting Node.js's crypto module with the Web Crypto API
> > Check it out, might be of help
> 
> [@arnavgupta00](https://github.com/arnavgupta00) Does your implementation works with any of [Next-Auth's built-in OAuth providers](https://next-auth.js.org/providers/)?
> 
> I have run into the same `crypto.createCipheriv is not implemented yet!` issue, but using [next-authv5](https://authjs.dev/getting-started/migrating-to-v5) and trying to authenticate with Google, as opposed to v4 from your example. It seems to me, in v5 at least, that `crypto.createCipheriv` is being called by the OAuth providers themselves.
> 
> Curious to know if there's a workaround here.

Hey @nathan-scheele, I ran into the same issue when trying to implement this with Google Auth in NextAuth v5. While I managed to get email OTP via Resend working, Google authentication remained unresolved. Spent a good amount of time trying to work around it but no luck so far. Let me know if you find a solution!

**Comment by contrun** [comment]
I am also facing this problem. Does any one know how to skip jwt verification in next-auth.js. I am doing a prototype for ȧ website. It is OK for me to just skill the verification all together. I don't want crypto.createCipheriv not implemented to block my other progresses.

**Comment by sassanh** [comment]
@contrun It is already implemented [here](https://github.com/cloudflare/workerd/pull/3635)
And was released yesterday as part of 2025-03-11 as you can see [here](https://github.com/cloudflare/workerd/compare/v1.20250310.0...v1.20250311.0#diff-774a9ae84685a34f9e292616eccd8a4fbd6735b5f92b20bb0e7a069192a8216dR419)

But unfortunately unenv is not yet updated with this change, I think it should be updated [here](https://github.com/cloudflare/workers-sdk/blob/869ec7b916487ec43b958a27bdfea13588c5685f/packages/unenv-preset/src/preset.ts). So unenv will report it is not implemented with this error:
```
✘ [ERROR] [auth][error] Error: [unenv] crypto.createCipheriv is not implemented yet!
```

I think it is the last step.

**Comment by sassanh** [comment]
> But unfortunately unenv is not yet updated with this change, I think it should be updated [here](https://github.com/cloudflare/workers-sdk/blob/869ec7b916487ec43b958a27bdfea13588c5685f/packages/unenv-preset/src/preset.ts).

[Here](https://github.com/cloudflare/workers-sdk/blob/869ec7b916487ec43b958a27bdfea13588c5685f/packages/unenv-preset/src/runtime/node/crypto.ts#L106) actually

**Comment by vicb** [comment]
workers-sdk will be updated once the feature has been released and is available in prod.

**Comment by jasnell** [comment]
Given that `crypto.createCipheriv` has been implemented I'm going to close this issue now. The workers-sdk update, as @vicb mentions should be coming very soon. If you run into any issues with the implementation, please either open a new issue (preferred) or comment here and mention me. Thanks all!

**Comment by vicb** [comment]
You will see that the unenv preset is updated in the wrangler changelog at https://developers.cloudflare.com/workers/platform/changelog/wrangler/

Sometimes later this week.

**Comment by imWildCat** [comment]
I see. Thank you very much!

**Comment by FocusAware** [comment]
It seems this much anticipated fix is there. Has the update been picked up by opennextJs automatically?

**Comment by vicb** [comment]
It should go to the wrangler 4.x release tomorrow Friday.

**Comment by FocusAware** [comment]
Well, I have tried and am still still seeing the same error:
 ```
 (error) [auth][error] CallbackRouteError: Read more at https://errors.authjs.dev#callbackrouteerror
  (error) [auth][cause]: Error: [unenv] crypto.createCipheriv is not implemented yet!
    at createNotImplementedError (worker.js:1:630)
    at fn (worker.js:1:729)
    at worker.js:3081:26475
    at eY (worker.js:3081:26601)
    at rs.encrypt (worker.js:3081:42131)
    at async wi.encrypt (worker.js:3081:42922)
    at async Object.tY [as encode] (worker.js:3081:56823)
    at async oU (worker.js:3458:49640)
    at async oM (worker.js:3458:59146)
    at async oJ (worker.js:3458:63661)
  (error) [auth][details]: {
  "provider": "guest"


npx wrangler --version

 ⛅️ wrangler 4.3.0
------------------

```

**Comment by imWildCat** [comment]
for those who have been tracking the progress, IMO wrangler 4.4.0 will be released once this PR is merged: https://github.com/cloudflare/workers-sdk/pull/8613
(I personally subscribed to this PR)

**Comment by penalosa** [comment]
@imWildCat You can try out the changes today by installing `wrangler@beta`, which should let you experiment before v4.4 goes out

**Comment by imWildCat** [comment]
thank you @penalosa !
after seeing your comment, I immediately tested my projects.
`crypto.createCipheriv` can be called.
But nextauth still throws error without any logs.

As of now I have to move back to Docker + Linux VM for our service. Will be back once it is stable and production ready.

**Comment by FocusAware** [comment]
Nextauth works for me now! Although I've tried my own JWT token based provider for guest sessions, but not others.
